python/rascaline/rascaline/utils/splines.py

```diff
@@ -0,0 +1,396 @@
+"""
+Splined radial integrals
+========================
+
+Classes for generating spline points which can be used as input for the for the radial
+integral hyper paramater option. For an complete example for the LE how to use the
+classes see :ref:`example-splines`.
+
+.. autoclass:: rascaline.utils.SplinerBase
+    :members:
+    :show-inheritance:
+
+.. autoclass:: rascaline.utils.SplinerFromFunction
+    :members:
+    :show-inheritance:
+"""
+
+from abc import ABC, abstractmethod
+from typing import Callable, Dict, Optional, Union
+
+import numpy as np
+
+
+class SplinerBase(ABC):
+    """Base class for splining arbitrary radial integrals.
+
+    If ``_radial_integral_derivative`` is not implemented in a child class it will
+    computed based on finite differences.
+
+    :parameter max_angular: number of radial components
+    :parameter max_radial: number of angular components
+    :parameter spline_cutoff: cutoff radius for the spline interpolation. This is also
+        the maximal value that can be interpolated.
+    :parameter accuracy: accuracy of the numerical integration and the splining.
+        Accuracy is reached when either the mean absolute error or the mean relative
+        error gets below the ``accuracy`` threshold.
+    """
+
+    def __init__(
+        self,
+        max_radial: int,
+        max_angular: int,
+        spline_cutoff: float,
+        accuracy: float,
+    ):
+        self.max_radial = max_radial
+        self.max_angular = max_angular
+        self.spline_cutoff = spline_cutoff
+        self.accuracy = accuracy
+
+    @abstractmethod
+    def _radial_integral(self, n: int, ell: int, positions: np.ndarray) -> np.ndarray:
+        """Method calculating the radial integral."""
+        ...
+
+    @property
+    def _center_contribution(self) -> Union[None, np.ndarray]:
+        r"""Contribution of the central atom required for LODE calculations.
+
+        The `center_contribution` is defined as :math:`c_n = \sqrt{4π} \int dr r^2
+        R_n(r) g(r)` where :math:`g(r)` is the radially symmetric density function,
+        `R_n(r)` the radial basis function and :math:`n` the current radial channel.
+        Note that the integration range was deliberately left ambiguous since it depends
+        on the radial basis, i.e. for the GTO basis, :math:`r \in R^+` is used, while
+        :math:`r \in [0, cutoff]` for the monomial basis.
```

Comment:
I still disagree with this formulation ^_^. this is always an integral to +\infty, but sometime the function to integrate is 0 outside of the cutoff, so the integral can be simplified. But we can always define it as an integral from 0 to +\infty.

Reply:
fine for me to change it.